diff --git a/node_modules/starlight-openapi/index.ts b/node_modules/starlight-openapi/index.ts
index bf83987..c61108f 100644
--- a/node_modules/starlight-openapi/index.ts
+++ b/node_modules/starlight-openapi/index.ts
@@ -31,7 +31,7 @@ export default function starlightOpenAPIPlugin(userConfig: StarlightOpenAPIUserC
           config.map((schemaConfig) => parseSchema(logger, astroConfig.root, schemaConfig)),
         )
 
-        addRouteMiddleware({ entrypoint: 'starlight-openapi/middleware', order: 'post' })
+        addRouteMiddleware({ entrypoint: 'starlight-openapi/middleware', order: 'pre' })
         addIntegration(starlightOpenAPIIntegration(starlightConfig, schemas))
 
         const updatedConfig: Parameters<typeof updateConfig>[0] = {
