diff --git a/example.env b/example-production.env
index c411265..8ffb74e 100644
--- a/example.env
+++ b/example-production.env
@@ -2,53 +2,54 @@
 # - - -
 
 # =============================================================================
-# DEVELOPMENT CONFIGURATION
+# PRODUCTION CONFIGURATION
+# Edit the values in this section when deploying a production instance.
+# For details, please visit https://candig.github.io/CanDIGv2/deployment/production/#env-settings
 # =============================================================================
 
-CANDIG_DOMAIN=candig.docker.internal
+CANDIG_DOMAIN=YOUR_PRODUCTION_DOMAIN
 CANDIG_INTERNAL_DOMAIN=${CANDIG_DOMAIN}
-CANDIG_AUTH_DOMAIN=candig.docker.internal
-CANDIG_SITE_LOCATION=local
-CANDIG_DEBUG_MODE=1
+CANDIG_AUTH_DOMAIN=YOUR_PRODUCTION_AUTH_DOMAIN
+CANDIG_SITE_LOCATION=YOUR_SITE_IDENTIFIER
+CANDIG_DEBUG_MODE=0
 LOCAL_IP_ADDR=
 
 # Federation
 FEDERATION_PORT=4232
 FEDERATION_SERVICE_URL=http://${CANDIG_INTERNAL_DOMAIN}:${FEDERATION_PORT}
-FEDERATION_SELF_SERVER_ID=internal-1
-FEDERATION_SELF_SERVER="{'id': '${FEDERATION_SELF_SERVER_ID}', 'url': '${FEDERATION_SERVICE_URL}','location': {'name': '${CANDIG_SITE_LOCATION}','province': 'ON','province-code': 'ca-on'}}"
+FEDERATION_SELF_SERVER_ID=YOUR-SERVER-ID
+FEDERATION_SELF_SERVER="{'id': '${FEDERATION_SELF_SERVER_ID}', 'url': '${FEDERATION_SERVICE_URL}','location': {'name': '${CANDIG_SITE_LOCATION}','province': 'YOUR-PROVINCE','province-code': 'YOUR-PROVINCE-CODE'}}"
 
 # Keycloak
-KEYCLOAK_PRODUCTION_MODE=0
-KEYCLOAK_CLIENT_ID=local_candig
+KEYCLOAK_PRODUCTION_MODE=1
+KEYCLOAK_CLIENT_ID=YOUR_KEYCLOAK_IDENTIFIER
 KEYCLOAK_PORT=8080
 KEYCLOAK_HTTPS_PORT=8443
-KEYCLOAK_PUBLIC_PROTO=http
+KEYCLOAK_PUBLIC_PROTO=https
 KEYCLOAK_PRIVATE_PROTO=http
-KEYCLOAK_PUBLIC_URL=${KEYCLOAK_PUBLIC_PROTO}://${CANDIG_AUTH_DOMAIN}:${KEYCLOAK_PORT}
-KEYCLOAK_PRIVATE_URL=${KEYCLOAK_PRIVATE_PROTO}://${CANDIG_AUTH_DOMAIN}:${KEYCLOAK_PORT}
-KEYCLOAK_PROXY_HEADERS=none
+KEYCLOAK_PUBLIC_URL=${KEYCLOAK_PUBLIC_PROTO}://${CANDIG_AUTH_DOMAIN}
+KEYCLOAK_PRIVATE_URL=${KEYCLOAK_PRIVATE_PROTO}://keycloak:${KEYCLOAK_PORT}
+KEYCLOAK_PROXY_HEADERS=xforwarded
 
 # Tyk gateway
-TYK_LOGIN_TARGET_URL=http://${CANDIG_DOMAIN}:${TYK_SERVICE_PUBLIC_PORT}
-TYK_ANALYTICS_FROM_EMAIL=admin@distributedgenomics.ca
-TYK_USE_SSL=false
+TYK_LOGIN_TARGET_URL=https://${CANDIG_DOMAIN}
+TYK_ANALYTICS_FROM_EMAIL=YOUR-ADMIN-EMAIL@email.ca
+TYK_USE_SSL=true
 
 # Data portal
 CANDIG_DATA_PORTAL_PORT=4173
 CANDIG_DATA_PORTAL_URL=https://${CANDIG_DOMAIN}:${CANDIG_DATA_PORTAL_PORT}/data-portal
-CANDIG_DATA_PORTAL_SUPPORT_EMAIL=site_admin@test.ca
+CANDIG_DATA_PORTAL_SUPPORT_EMAIL=YOUR_SITE_ADMIN_SUPPORT_EMAIL
 
 DIR_PERMISSIONS=775
 
 # =============================================================================
-# END DEVELOPMENT CONFIGURATION
+# END PRODUCTION CONFIGURATION
 # =============================================================================
 
 
 # site options
 CANDIG_MODULES=logging keycloak vault redis postgres drs htsget katsu rnaget query tyk opa candig-ingest federation candig-data-portal
-    #monitoring
 CANDIG_AUTH_MODULES=keycloak vault tyk opa federation
 CANDIG_DATA_MODULES=keycloak vault redis postgres logging
 
@@ -69,7 +70,7 @@ TEST_USER_2=user2@${CANDIG_SITE_LOCATION}-test.ca
 # Set KEEP_TEST_DATA to true to have the integration tests keep all test data
 KEEP_TEST_DATA=false
 
-# miniforge venv
+# miniconda venv
 # options are [linux, darwin, arm64mac]
 VENV_OS=linux
 VENV_NAME=candig
@@ -286,7 +287,7 @@ PROGRESS_FILE=${PWD}/tmp/progress.txt
 # install logging
 LOGFILE=tmp/progress.txt
 
-CONDA_INSTALL=bin/miniforge
+CONDA_INSTALL=bin/miniconda3
 
 COMPOSE_IGNORE_ORPHANS=True
 
@@ -294,6 +295,7 @@ COMPOSE_IGNORE_ORPHANS=True
 # Options are ROPC CLIENT AUTH_CODE
 OIDC_CHAIN=ROPC
 # AUTH_CODE-specific variables (for use in auth_code_acceptor.py):
+# NB: The AUTH_ACCEPT_URL might not work in most cases -- please configure your reverse proxy accordingly
 AUTH_ACCEPT_PORT=8989
 AUTH_ACCEPT_URL=http://${CANDIG_DOMAIN}:${AUTH_ACCEPT_PORT}
 # CLIENT-specific variables
