diff --git a/example.env b/example-production.env
index 64faf82..42b6a9b 100644
--- a/example.env
+++ b/example-production.env
@@ -2,14 +2,16 @@
 # - - -
 
 # =============================================================================
-# DEVELOPMENT CONFIGURATION
+# PRODUCTION CONFIGURATION
+# Edit the values in this section when deploying a production instance.
+# For details, please visit https://candig.github.io/CanDIGv2/deployment/production/#env-settings
 # =============================================================================
 
-CANDIG_DOMAIN=candig.docker.internal
+CANDIG_DOMAIN=YOUR_PRODUCTION_DOMAIN
 CANDIG_INTERNAL_DOMAIN=${CANDIG_DOMAIN}
-CANDIG_AUTH_DOMAIN=candig.docker.internal
-CANDIG_SITE_LOCATION=local
-CANDIG_DEBUG_MODE=1
+CANDIG_AUTH_DOMAIN=YOUR_PRODUCTION_AUTH_DOMAIN
+CANDIG_SITE_LOCATION=YOUR_SITE_IDENTIFIER
+CANDIG_DEBUG_MODE=0
 LOCAL_IP_ADDR=
 # miniforge venv
 # options are [linux, darwin, arm64mac]
@@ -22,38 +24,36 @@ KEEP_TEST_DATA=false
 # Federation
 FEDERATION_PORT=4232
 FEDERATION_SERVICE_URL=http://${CANDIG_INTERNAL_DOMAIN}:${FEDERATION_PORT}
-FEDERATION_SELF_SERVER_ID=internal-1
-FEDERATION_SELF_SERVER="{'id': '${FEDERATION_SELF_SERVER_ID}', 'url': '${FEDERATION_SERVICE_URL}','location': {'name': '${CANDIG_SITE_LOCATION}','province': 'ON','province-code': 'ca-on'}}"
+FEDERATION_SELF_SERVER_ID=YOUR-SERVER-ID
+FEDERATION_SELF_SERVER="{'id': '${FEDERATION_SELF_SERVER_ID}', 'url': '${FEDERATION_SERVICE_URL}','location': {'name': '${CANDIG_SITE_LOCATION}','province': 'YOUR-PROVINCE','province-code': 'YOUR-PROVINCE-CODE'}}"
 
 # Keycloak
-KEYCLOAK_PRODUCTION_MODE=0
-KEYCLOAK_CLIENT_ID=local_candig
+KEYCLOAK_PRODUCTION_MODE=1
+KEYCLOAK_CLIENT_ID=YOUR_KEYCLOAK_IDENTIFIER
 KEYCLOAK_PORT=8080
 KEYCLOAK_HTTPS_PORT=8443
-KEYCLOAK_PUBLIC_PROTO=http
+KEYCLOAK_PUBLIC_PROTO=https
 KEYCLOAK_PRIVATE_PROTO=http
-KEYCLOAK_PUBLIC_URL=${KEYCLOAK_PUBLIC_PROTO}://${CANDIG_AUTH_DOMAIN}:${KEYCLOAK_PORT}
-KEYCLOAK_PRIVATE_URL=${KEYCLOAK_PRIVATE_PROTO}://${CANDIG_AUTH_DOMAIN}:${KEYCLOAK_PORT}
-KEYCLOAK_PROXY_HEADERS=none
+KEYCLOAK_PUBLIC_URL=${KEYCLOAK_PUBLIC_PROTO}://${CANDIG_AUTH_DOMAIN}
+KEYCLOAK_PRIVATE_URL=${KEYCLOAK_PRIVATE_PROTO}://keycloak:${KEYCLOAK_PORT}
+KEYCLOAK_PROXY_HEADERS=xforwarded
 
 # Tyk gateway
-TYK_LOGIN_TARGET_URL=http://${CANDIG_DOMAIN}:${TYK_SERVICE_PUBLIC_PORT}
-TYK_ANALYTICS_FROM_EMAIL=admin@distributedgenomics.ca
-TYK_USE_SSL=false
+TYK_LOGIN_TARGET_URL=https://${CANDIG_DOMAIN}
+TYK_ANALYTICS_FROM_EMAIL=YOUR-ADMIN-EMAIL@email.ca
+TYK_USE_SSL=true
 
 # Data portal
 CANDIG_DATA_PORTAL_PORT=4173
 CANDIG_DATA_PORTAL_URL=https://${CANDIG_DOMAIN}:${CANDIG_DATA_PORTAL_PORT}/data-portal
-CANDIG_DATA_PORTAL_SUPPORT_EMAIL=site_admin@test.ca
+CANDIG_DATA_PORTAL_SUPPORT_EMAIL=YOUR_SITE_ADMIN_SUPPORT_EMAIL
 MOHCCN_SUPPORT_EMAIL=mohdatarequests@tfri.ca
 
 # =============================================================================
-# END DEVELOPMENT CONFIGURATION
-# Variables in this section are unlikely to need to be changed. 
-# Please consult with a CanDIG developer before changing if you aren't 
-# familiar with what the variable does.
+# END PRODUCTION CONFIGURATION
 # =============================================================================
 
+
 # site options
 CANDIG_MODULES=logging keycloak vault redis postgres drs htsget katsu rnaget query tyk opa candig-ingest federation candig-data-portal
 CANDIG_AUTH_MODULES=keycloak vault tyk opa federation
