{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A Brief Primer on Variant Calling\n",
    "\n",
    "The DNA sequences of two individuals in a population will be different because of crossover between parents' genes and because of random mutations. For a wide variety of applications, it has become important to know an individual's genetic composition (or genotype). But do we need to store the whole sequence?\n",
    "\n",
    "It turns out that, instead of storing the whole sequence, we can simply store the difference between an individual's DNA sequence and a reference individual, as illustrated in this image from the [European Bioinformatics Institute](https://www.ebi.ac.uk/training/online/course/human-genetic-variation-i-introduction/variant-identification-and-analysis/what-variant):\n",
    "![Generate Variants](https://www.ebi.ac.uk/training/online/sites/ebi.ac.uk.training.online/files/resize/GenVar_Fig_CRAM_file-750x162.png)\n",
    "\n",
    "In the image, the gray lines are \"reads\" from the sequencing instrument. The red letters indicate the areas where the reads differ from the reference DNA sequence (depicted in color). These differences are called variants, and the variants are stored in a tab-delimited text file called a Variant Caller Format (VCF) file.\n",
    "\n",
    "Identifying variants requires a sophisticated algorithm such as GATK or DeepVariant. For one thing, the reads from the instrument have to be aligned to the reference DNA sequence. The reads may not exactly match any part of the reference because ... variants. Even though the image above shows differences in just one letter (or single nucleotide polymorphism), variants can be more complex -- entire subsequences might be deleted, inserted, inverted, duplicated, or duplicated repeatedly:\n",
    "\n",
    "![Types of Variants](https://codelabs.developers.google.com/codelabs/genomics-deepvariant/img/f7142c3c5224fe9c.png)\n",
    "\n",
    "*Image from European Bioinformatics Institute.*\n",
    "\n",
    "Variants are significant, not just because they allow more efficient storage, but because genetic differences between individuals can lead to differences in an individual's phenotype (such as their height or risk of developing a disease).\n",
    "\n",
    "Variant calling is the process of going from sequences (typically stored in a BAMS files) to identifying variants between an individual and a reference genome (typically stored in a VCF file). Once we have the variants, then it is possible to do analysis on the variants to look for specific genetic markers.\n",
    "\n",
    "**Test your understanding:**\n",
    "\n",
    "    DeepVariant is an algorithm to do _________________\n",
    "    There are _______ inputs to a variant caller.\n",
    "    The inputs to a variant caller are _______________ and _________________.\n",
    "    The output of a variant caller is __________________\n",
    "\n",
    "**Answers to \"Test your understanding\":**\n",
    "\n",
    "    Variant calling.\n",
    "    Two. (see below)\n",
    "    A reference genome and sequence reads belonging to an individual tissue. The reference could belong to the species or to the same individual organism (such as from the organism a few years prior, or from a part of the body unaffected by disease).\n",
    "    A VCF file containing the variants between the reference genome and the individual tissue's genotype.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DeepVariant\n",
    "\n",
    "DeepVariant is an analysis pipeline that uses a deep neural network to call genetic variants from next-generation DNA sequencing data. It is a TensorFlow-based image classification model ([inception-v3](https://arxiv.org/abs/1512.00567)) to assign genotype likelihoods from the experimental data produced by the instrument. Read additional information on the [Google Research blog.](https://ai.googleblog.com/2017/12/deepvariant-highly-accurate-genomes.html)\n",
    "\n",
    "![DeepVariant Workflow](https://github.com/google/deepvariant/blob/r0.7/docs/DeepVariant-workflow-figure.png?raw=true)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using DeepVariant to Build Analyze Variants\n",
    "\n",
    "These portable WDL workflows use DeepVariant to call variants from WGS read alignments, followed by GLnexus to merge the resulting Genome VCF (gVCF) files for several samples into a Project VCF (pVCF). This sequential workflow to generate gVCF from a given BAM file and genomic range.\n",
    "\n",
    "```\n",
    "             +----------------------------------------------------------------------------+\n",
    "             |                                                                            |\n",
    "             |  DeepVariant.wdl                                                           |\n",
    "             |                                                                            |\n",
    "             |  +-----------------+    +-----------------+    +------------------------+  |\n",
    "sample.bam   |  |                 |    |                 |    |                        |  |\n",
    " genome.fa ----->  make_examples  |---->  call_variants  |---->  postprocess_variants  |-----> gVCF\n",
    "     range   |  |                 |    |                 |    |                        |  |\n",
    "             |  +-----------------+    +--------^--------+    +------------------------+  |\n",
    "             |                                  |                                         |\n",
    "             |                                  |                                         |\n",
    "             +----------------------------------|-----------------------------------------+\n",
    "                                                |\n",
    "                                       DeepVariant Model\n",
    "```\n",
    "\n",
    "`make_examples` and `call_variants` internally parallelize across CPUs on the machine they run on. The tasks use the docker image published by the [DeepVariant](https://github.com/google/deepvariant) team.\n",
    "\n",
    "## Htsget + Deepvariant + WDL\n",
    "For each range, fetches a BAM slice using the GA4GH htsget client in samtools 1.7+, given an htsget server endpoint and sample ID. Finally, concatenates the per-range gVCFs to the complete product.\n",
    "\n",
    "```\n",
    "             +--------------------------------------------------------------------------------+\n",
    "             |                                                                                |\n",
    "             |  htsget_DeepVariant.wdl                                                        |\n",
    "             |                                                                                |\n",
    "             |       +-----------------+    +-------------------+                             |\n",
    "             |       |                 |    |                   |  range gVCF                 |\n",
    "             |   +--->  htsget client  |---->  DeepVariant.wdl  |---+                         |\n",
    "             |   |   |  (samtools)     |    |                   |   |                         |\n",
    "             |   |   |                 |    +-------------------+   |                         |\n",
    "sample ID    |   |   +-----------------+                            |  +-------------------+  |\n",
    "             |   |                                                  +-->                   |  |\n",
    "   ranges -------+---> ...                  ...                 ... --->  bcftools concat  +-----> sample gVCF\n",
    "    (e.g.    |   |                                                  +-->                   |  |\n",
    "     chr1    |   |   +-----------------+                            |  +-------------------+  |\n",
    "     chr2    |   |   |                 |    +-------------------+   |                         |\n",
    "     ...)    |   +--->  htsget client  |    |                   |   |                         |\n",
    "             |       |  (samtools)     |---->  DeepVariant.wdl  |---+                         |\n",
    "             |       |                 |    |                   |  range gVCF                 |\n",
    "             |       +------------^----+    +-------------------+                             |\n",
    "             |            |       |                                                           |\n",
    "             |            |       |                                                           |\n",
    "             +------------|-------|-----------------------------------------------------------+\n",
    "                          |       |\n",
    "               sample ID  |       |\n",
    "                   range  |       |  range BAM\n",
    "                          |       |\n",
    "                     +----v------------+\n",
    "                     |                 |\n",
    "                     |  htsget server  |\n",
    "                     |                 |\n",
    "                     +-----------------+\n",
    "```\n",
    "\n",
    "By using htsget, the workflow scatters across the ranges without first having to download and slice up a monolithic BAM file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}

